website/addons/onedrive/model.py

```diff
@@ -0,0 +1,240 @@
+        :param str refresh_token: Refresh token used to request a new access token
+        :return dict token: New set of tokens
+        """
+        client = self._auth_client
+        if refresh_token:
+            token = client.refresh(access_token, refresh_token)
+            return token
+        else:
+            return False
+
+    def fetch_access_token(self, force_refresh=False):
+        self.refresh_access_token(force=force_refresh)
+        return self.account.oauth_key
+
+    def refresh_access_token(self, force=False):
+        """ If the token has expired or will soon, handles refreshing and the storage of new tokens
+
+        :param bool force: Indicates whether or not to force the refreshing process, for the purpose of ensuring that authorization has not been unexpectedly removed.
+        """
+        if self._needs_refresh() or force:
+            token = self._refresh_token(self.account.oauth_key, self.account.refresh_token)
+            self.account.oauth_key = token['access_token']
+            self.account.refresh_token = token['refresh_token']
+            self.account.expires_at = datetime.utcfromtimestamp(token['expires_at'])
+            self.account.save()
+
+    def _needs_refresh(self):
+        if self.account.expires_at is None:
+            return False
+        return (self.account.expires_at - datetime.utcnow()).total_seconds() < settings.REFRESH_TIME
+
+class OneDriveUserSettings(AddonOAuthUserSettingsBase):
+    """Stores user-specific onedrive information
+    """
+    oauth_provider = OneDrive
+    serializer = OneDriveSerializer
+
+
+class OneDriveNodeSettings(StorageAddonBase, AddonOAuthNodeSettingsBase):
+
+    oauth_provider = OneDrive
+    serializer = OneDriveSerializer
+
+    foreign_user_settings = fields.ForeignField(
+        'onedriveusersettings', backref='authorized'
+    )
+    folder_id = fields.StringField(default=None)
+    onedrive_id = fields.StringField(default=None)
+    folder_name = fields.StringField()
+    folder_path = fields.StringField()
+
+    _folder_data = None
+
+    _api = None
+
+    @property
+    def api(self):
+        """authenticated ExternalProvider instance"""
+        if self._api is None:
+            self._api = OneDrive(self.external_account)
+        return self._api
+
+    @property
+    def display_name(self):
+        return '{0}: {1}'.format(self.config.full_name, self.folder_name)
+
+    @property
+    def has_auth(self):
+        """Whether an access token is associated with this node."""
+        return bool(self.user_settings and self.user_settings.has_auth)
+
+    @property
+    def complete(self):
+        return bool(self.has_auth and self.user_settings.verify_oauth_access(
+            node=self.owner,
+            external_account=self.external_account,
+        ))
+
+    def fetch_folder_name(self):
+        self._update_folder_data()
+        return self.folder_name.replace('All Files', '/ (Full OneDrive)')
```

Comment:
Disregard, saw that it was being set as such in the view (presently, at least). I think this function, as well as `fetch_full_folder_path` and `_update_folder_data`, can be removed. Any references to them should be using `self.folder_name`